### ISSUE TITLE
Question on the Summary Tool

### ISSUE THREAD

**Issue by TomLucidor** [cognitivetech/ollama-ebook-summary#5]
1. How much support does this have for PDF, EPUB, MOBI, and MarkDown?
2. Why only 2000 Token chunks when most LLMs can handle easily 4000+ Tokens? (yes "forgotten middle" is an issue for longer context)
3. Are there advantages to break the book based on chapters or topics (and paragraphs) instead of arbitrary token count?
4. What other models are out there that are best suited for either text embedding or summarization? Are finetunes any better, and if so what would the tuning methodology be?
5. Are there benchmarks that makes quality determination of the newer LLMs easier? Everything from Llama finetunes to Mistral and Nemotron and Command R are getting recommended online

**Comment by cognitivetech** [comment]
1. currently cover some pdf, most epub, and markdown is markdown.. can choose what level to split by

2. chunk size was chosen after 6 months of research and testing, these are the best quality of results at that size  (you can look in [notes/depreciated/](https://github.com/cognitivetech/ollama-ebook-summary/tree/main/notes/depreciated) for more details.)

that doesn't have to remain forever, but how the current fine-tune is trained. I'm open to experimenting with more, but not much.. 
I already know when choosing higher chunk size:
- quality goes down
- level of detail in response is reduced
- hallucinations go up (even for SOTA)

3. yes absolutely advantage to split by section for a more high quality production grade output.

if you look at the outputs I create markdown file following the original ToC... if you daon't care about how to navigate and how it was originally structured, then do whatever you are into.

4. There is none, that's why I made this. 

5. None of those new models have done better out of box than what mistral 7b v0.2 did (for the demands I apply to this particular task, i don't care what your benchmarks said they don't follow basic instructions is why I finetune)

incidentally, I am using llm for creating titles now, but I will move back to something more "traditional" ml because its more quality and performant for that specific task.

6. That's quite beyond the scope of this project and even logistically the idea of having books argue with eachother would require to fine-tune models specifically on that entirety of book.. 

in silly-tavern its possible to create entire world-books, and you could set that up based on outputs created from this tool, possibly... but again, way outside of the scope.

P.S. That mapreduce stuff is interesting. I have my own plans to do something simiar, so eventually I'll have a few levels of specificity...

I've done my due dilligence on this project and created this tool because there was no other project which really addressed the subject... and when I began it was during the long-context craze where people insist ont making "long-context" models that no matter how big or small of context provided, they could not reliably following instructions to produce a consistent output)

**Comment by TomLucidor** [comment]
More on No.4, I think a casual browse of the internet would recommend models like (and I mean people recommend a lot of things): Tess, Llama2 or Llama3, Nous-Capybara, Pegasus,  Command R+, Mistral SciPhi, Mixtral, DBRX, Artic, Tinyllama, Starling, NoroCetacean, Miqu, led-base-book-summary, long-t5, distilbart, Inkblot, Yi, una-cybertron, phi-2 or phi-3, OpenOrca, GLM, Sensei, Gemma, Qwen2...
Yeah they are all of different sizes so each weight class might need their own evaluations

**Comment by cognitivetech** [comment]
that's an interesting idea to create benchmarks for summary. but that's a very opinionated task I'm afraid.

I tried hundreds of models myself, including most of those you mentioned.

its a lot of work.

Better just find one that works, or that you can make work, and move forward.

The existing fine-tune I have could be improved...

that's on my plan, I already know I can make a better dataset with llama3 405b, which in turn can make a better fine-tune.

but according to what I want NONE of the models off-shelf will fit on my GPU and do a better job than the fine-tune I have.

that's based on my experience for the last year of FOMO downloading every model and thinking "OMG it has great benchmarks it will do great" and finding they did not.

**Comment by cognitivetech** [comment]
to be clear, when I say "Benchmarks would be a highly opinionated task" what I mean is that whatever tests and benchmarks I looked at involved summary of very small text. They only ask for summary. This tool is not just a summary tool. 

I create comprehensive bulleted notes. I care about format. I'm summarizing entire books and want to be able to quickly navigate and review the information.

So I chose the models best able to follow the format, and presume that if they follow instructions better, they understand language better, and they will produce better summaries.

Ultimately, I would like to create benchmarks to test various models, that's how this project began, idk if you browsed the notes I shared earlier... 

before I released a package I did 7 different tests of what were purported to be the top models of that time, and they mostly did a miserable job at **consistently** creating the quality I asked for.

So I created a fine-tune of Mistral 7b v0.3 Instruct, using a dataset I created with the same (manually polished) after finding the best template to get 90% quality of results I was looking for.

Since that time, I've tried all the best local models which will fit on my gpu. All the big ones anyways. 

Looking better in the 70b range than at the time I did my original testing. I can get at least one of those to work with just a prompt. and as I said llama3 405b does a good job at following instructions and is linguistically comparable to Opus.

However, I'm not sure I would choose llama3 8b for fine-tuning. 

I will probably do my next fine tune, creating summaries with 405b, and use one of the new mistral models for the tune. Maybe llm38b is a good choice too, I've had some good conversations with it..  but I heard that it is more difficult to train.

Semantic similarity would be relatively simple type of benchmark, and there is probably research and code demonstrating that. I really haven't looked at the research on this specific topic, but I will do that before proceeding to that stage.

Otherwise, I am only more recently following the research, so I've only begun to see what people are doing with summaries.

**Comment by TomLucidor** [comment]
@cognitivetech Thanks for the explanation, regarding bullet points and formatting I totally agree in this format (but also slightly biased towards having another format like having one-line summaries for sections), and also methods like Chain of Density for capturing "key ideas" that people might miss. Maybe there is an element of prompt engineering for models?
But at the same time I do have to note that there are a lot of different benchmark on testing summarization, and yes semantic similarity is one of them but I am not sure what other method or metric could make sense...

**Comment by cognitivetech** [comment]
if you need another type of format, then I just send that prompt to a general purpose model. "Provide a sentence summarizing this text"

as far as benchmarks, I will get there someday.

this is a really big project, there are many aspects which make it challenging. So I'm just driving forward... right now I have a great tool I've been using on the daily for maybe 9 months and is quite usable (for school and work)

there are more features I'm focused on including with pdf how to split sections between paragraphs.. as that just takes you to a page, not a specific location

I do want to fine-tune a new model here in the coming months, as current version works fine, but we can do a lot better now.

once I've manually calibrated to optimum ideals, then more tests will be useful.
